utils/cautils/tpm.go

```diff
@@ -0,0 +1,651 @@
+package cautils
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/google/go-attestation/attest"
+	"github.com/urfave/cli"
+
+	"github.com/smallstep/certificates/acme"
+	"github.com/smallstep/certificates/ca"
+	"github.com/smallstep/cli/utils"
+	"go.step.sm/cli-utils/ui"
+	"go.step.sm/crypto/jose"
+	"go.step.sm/crypto/keyutil"
+	"go.step.sm/crypto/kms/uri"
+	"go.step.sm/crypto/tpm"
+	tpmstorage "go.step.sm/crypto/tpm/storage"
+)
+
+func doTPMAttestation(clictx *cli.Context, ac *ca.ACMEClient, ch *acme.Challenge, identifier string, af *acmeFlow) error {
+	tpmStorageDirectory := clictx.String("tpm-storage-directory")
+	t, err := tpm.New(tpm.WithStore(tpmstorage.NewDirstore(tpmStorageDirectory)))
+	if err != nil {
+		return fmt.Errorf("failed initializing TPM: %w", err)
+	}
+
+	tpmAttestationCABaseURL := clictx.String("attestation-ca-url")
+	if tpmAttestationCABaseURL == "" {
+		return fmt.Errorf("flag %q cannot be empty", "--attestation-ca-url")
+	}
+
+	tpmAttestationCARootFile := clictx.String("attestation-ca-root")
+	tpmAttestationCAInsecure := clictx.Bool("attestation-ca-insecure")
+
+	insecure := clictx.Bool("insecure")
+	kty, crv, size, err := utils.GetKeyDetailsFromCLI(clictx, insecure, "kty", "curve", "size")
+	if err != nil {
+		return fmt.Errorf("failed getting key details: %w", err)
+	}
+
+	var inputSize int
+	inputKeyType := kty
+	switch kty {
+	case "EC":
+		switch crv {
+		case "P-256":
+			inputSize = 256
+		case "P-384":
+			inputSize = 384
+		case "P-521":
+			inputSize = 521
+		default:
+			return fmt.Errorf("unsupported curve: %q", crv)
+		}
+		inputKeyType = "ECDSA"
+	case "RSA":
+		inputSize = size
+	default:
+		return fmt.Errorf("unsupported key type: %q", kty)
+	}
+
+	attestationURI := clictx.String("attestation-uri")
+	keyName, err := parseTPMAttestationURI(attestationURI)
+	if err != nil {
+		return fmt.Errorf("failed parsing --attestation-uri: %w", err)
+	}
+
+	ui.Printf("Using Device Attestation challenge to validate %q", identifier)
+	ui.Printf(" .") // Indicates passage of time.
+
+	ctx := tpm.NewContext(context.Background(), t)
+	info, err := t.Info(ctx)
+	if err != nil {
+		return fmt.Errorf("failed retrieving TPM info: %w", err)
+	}
+
+	ui.Printf("\nTPM INFO:")
+	ui.Printf("\nVersion: %s", info.Version)
+	ui.Printf("\nInterface: %s", info.Interface)
+	ui.Printf("\nManufacturer: %s", info.Manufacturer)
+	ui.Printf("\nVendor info: %s", info.VendorInfo)
+	ui.Printf("\nFirmware version: %s", info.FirmwareVersion)
```

Comment:
I've changed the order of the output, so that the TPM info comes first. The current output is kind of in line with the ACME status updates:

```console
✔ Provisioner: attestation (ACME)
TPM INFO:
Version: TPM 2.0
Interface: kernel-managed
Manufacturer: Infineon (IFX, 49465800, 1229346816)
Vendor info: SLB9670
Firmware version: 7.85
Using Device Attestation challenge to validate "localhost" . done!
Waiting for Order to be 'ready' for finalization .. done!
Finalizing Order .. done!
✔ Would you like to overwrite l.crt [y/n]: y
✔ Certificate: l.crt
✔ Private Key: tpmkms:name=newkey2
```

I think it's useful to have this output at this stage for support/debugging. I've been thinking a bit about a `--verbose` option for more logging like this. But I think it's OK to add that later (and to more commands).